Issue resolving a method call with a variadic argument of primitive type

Given a code like `IntStream.of(values).map(s -> s).average()`, I'd like to resolve the type of `IntStream.of(values)` to infer it is an IntStream.

Related source is:

    public static IntStream of(int t) {
        return StreamSupport.intStream(new Streams.IntStreamBuilderImpl(t), false);
    }

    public static IntStream of(int... values) {
        return Arrays.stream(values);
    }

I encounter unexpected exception during the resolution:

```
Exception in thread "main" UnsolvedSymbolException{context='IntStream.of(values).map(s -> s)', name='IntStream.of(values)', cause='java.lang.UnsupportedOperationException: Unable to resolve the type typeParametersValues in a MethodUsage. Expected type: PrimitiveTypeUsage{name='int'}, Actual type: ResolvedArrayType{PrimitiveTypeUsage{name='int'}}. Method Declaration: ReflectionMethodDeclaration{method=public static java.util.stream.IntStream java.util.stream.IntStream.of(int[])}. MethodUsage: MethodUsage{declaration=ReflectionMethodDeclaration{method=public static java.util.stream.IntStream java.util.stream.IntStream.of(int[])}, paramTypes=[ResolvedArrayType{PrimitiveTypeUsage{name='int'}}]}'}
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractJavaParserContext.findTypeDeclarations(AbstractJavaParserContext.java:208)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveMethod(MethodCallExprContext.java:149)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:273)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:134)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:461)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:65)
	at com.github.javaparser.ast.expr.LambdaExpr.accept(LambdaExpr.java:164)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:503)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:380)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:651)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:323)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:65)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:118)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:503)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:350)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:332)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveMethodAsUsage(MethodCallExprContext.java:101)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:659)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:323)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:65)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:118)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:503)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:350)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:332)
	at com.github.javaparser.symbolsolver.JavaSymbolSolver.calculateType(JavaSymbolSolver.java:294)
	at com.github.javaparser.ast.expr.Expression.calculateResolvedType(Expression.java:575)
	at eu.solven.cleanthat.engine.java.refactorer.report_javaparser.ReportResolveTypeIntStreamOfIntArray.main(ReportResolveTypeIntStreamOfIntArray.java:53)
Caused by: java.lang.UnsupportedOperationException: Unable to resolve the type typeParametersValues in a MethodUsage. Expected type: PrimitiveTypeUsage{name='int'}, Actual type: ResolvedArrayType{PrimitiveTypeUsage{name='int'}}. Method Declaration: ReflectionMethodDeclaration{method=public static java.util.stream.IntStream java.util.stream.IntStream.of(int[])}. MethodUsage: MethodUsage{declaration=ReflectionMethodDeclaration{method=public static java.util.stream.IntStream java.util.stream.IntStream.of(int[])}, paramTypes=[ResolvedArrayType{PrimitiveTypeUsage{name='int'}}]}
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.resolveMethodTypeParameters(MethodCallExprContext.java:358)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveMethodAsUsage(MethodCallExprContext.java:91)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:659)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:323)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:65)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:118)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:503)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:350)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:332)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractJavaParserContext.findTypeDeclarations(AbstractJavaParserContext.java:199)
	... 26 more

```

Here is a minimal reproduction scenario:

```
import com.github.javaparser.JavaParser;
import com.github.javaparser.ParserConfiguration;
import com.github.javaparser.ast.expr.MethodCallExpr;
import com.github.javaparser.symbolsolver.JavaSymbolSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;

public class ReportResolveTypeIntStreamOfIntArray {
	static final String testCase = "import java.util.OptionalDouble;\n" + "import java.util.stream.IntStream;\n"
			+ "\n"
			+ "public class CaseIntStream {\n"
			+ "	OptionalDouble pre(int[] values) {\n"
			+ "		return IntStream.of(values).map(s -> s).average();\n"
			+ "	}\n"
			+ "}";

	public static ReflectionTypeSolver makeDefaultTypeSolver(boolean jreOnly) {
		var reflectionTypeSolver = new ReflectionTypeSolver(jreOnly);
		return reflectionTypeSolver;
	}

	public static JavaParser makeDefaultJavaParser(boolean jreOnly) {
		var reflectionTypeSolver = makeDefaultTypeSolver(jreOnly);

		var symbolResolver = new JavaSymbolSolver(reflectionTypeSolver);

		var configuration = new ParserConfiguration().setSymbolResolver(symbolResolver);
		var parser = new JavaParser(configuration);
		return parser;
	}

	public static void main(String[] args) {
		var node = makeDefaultJavaParser(true).parse(testCase).getResult().get();

		var ofCall = node.findAll(MethodCallExpr.class).get(0);

		System.out.println(ofCall.calculateResolvedType());
	}
}

```

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/MethodCallExprContext.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/MethodCallExprContext.java
@@ -340,13 +340,24 @@ public class MethodCallExprContext extends AbstractJavaParserContext<MethodCallE
                 // for example
                 // Arrays.aslist(int[]{1}) must returns List<int[]>
                 // but Arrays.aslist(String[]{""}) must returns List<String>
-                // May be the result depends on the component type of the array
+                // Arrays.asList() accept generic type T. Since Java generics work only on
+				// reference types (object types), not on primitives, and int[] is an object
+                // then Arrays.aslist(int[]{1}) returns List<int[]>
                 ResolvedType lastActualParamType =
                         actualParamTypes.get(actualParamTypes.size() - 1);
-                ResolvedType actualType =
-                		lastActualParamType.isArray() && lastActualParamType.asArrayType().getComponentType().isReferenceType() ?
-                				lastActualParamType.asArrayType().getComponentType() :
-                					lastActualParamType;
+                ResolvedType actualType = lastActualParamType;
+                if (lastActualParamType.isArray()) {
+                	ResolvedType componentType = lastActualParamType.asArrayType().getComponentType();
+                	// in cases where, the expected type is a generic type (Arrays.asList(T... a)) and the component type of the array type is a reference type
+                	// or the expected type is not a generic (IntStream.of(int... values)) and the component type is not a reference type
+                	// then the actual type is the component type (in the example above 'int')
+                	if ((componentType.isReferenceType()
+                			&& ResolvedTypeVariable.class.isInstance(expectedType))
+                			|| (!componentType.isReferenceType()
+                        			&& !ResolvedTypeVariable.class.isInstance(expectedType))) {
+                		actualType = lastActualParamType.asArrayType().getComponentType();
+                	}
+                }
                 if (!expectedType.isAssignableBy(actualType)) {
                     for (ResolvedTypeParameterDeclaration tp : methodUsage.getDeclaration().getTypeParameters()) {
                         expectedType = MethodResolutionLogic.replaceTypeParam(expectedType, tp, typeSolver);

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/JavaParserFacadeResolutionTest.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/JavaParserFacadeResolutionTest.java
@@ -250,6 +250,49 @@ class JavaParserFacadeResolutionTest extends AbstractResolutionTest {
         assertEquals("java.util.List<int[]>", resolvedType.describe());
     }
 
+    @Test
+    void resolveTypeParameterFromReferenceArrayArgument() {
+        String sourceCode = "" +
+                "import java.util.Arrays;\n" +
+                "\n" +
+                "public class Main {\n" +
+                "    public void main(String[] args) {\n" +
+                "        Arrays.asList(args);\n" +
+                "    }\n" +
+                "}";
+
+        JavaParser parser = createParserWithResolver(defaultTypeSolver());
+        CompilationUnit cu = parser.parse(sourceCode).getResult().get();
+
+        MethodCallExpr mce = cu.findFirst(MethodCallExpr.class).get();
+
+        ResolvedType resolvedType = mce.calculateResolvedType();
+
+        assertEquals("java.util.List<java.lang.String>", resolvedType.describe());
+    }
+
+    @Test
+    void resolveTypeParameterFromPrimitiveArrayArgumentOnNonGenericExpectedParameter() {
+        String sourceCode = "" +
+        		"import java.util.OptionalDouble;\n" +
+				"import java.util.stream.IntStream;\n" +
+                "\n" +
+                "public class Main {\n" +
+                "	OptionalDouble pre(int[] values) {\n" +
+				"		return IntStream.of(values).map(s -> s).average();\n" +
+				"	}\n" +
+                "}";
+
+        JavaParser parser = createParserWithResolver(defaultTypeSolver());
+        CompilationUnit cu = parser.parse(sourceCode).getResult().get();
+
+        MethodCallExpr mce = cu.findFirst(MethodCallExpr.class).get();
+
+        ResolvedType resolvedType = mce.calculateResolvedType();
+
+        assertEquals("java.util.OptionalDouble", resolvedType.describe());
+    }
+
     // See issue 3725
     @Test
     void resolveVarTypeInForEachLoopFromIterableExpression2() {

